subprojects/core/src/main/java/org/gradle/execution/commandline/CommandLineTaskConfigurer.java

```diff
@@ -57,16 +53,15 @@ private List<String> configureTasksNow(Collection<Task> tasks, List<String> argu
         List<String> remainingArguments = null;
         for (Task task : tasks) {
             CommandLineParser parser = new CommandLineParser();
-            Map<Boolean, List<OptionDescriptor>> allCommandLineOptions = optionReader.getOptions(task, false).stream().collect(Collectors.groupingBy(OptionDescriptor::isClashing));
-            List<OptionDescriptor> validCommandLineOptions = allCommandLineOptions.getOrDefault(false, Collections.emptyList());
-            List<OptionDescriptor> clashingOptions = allCommandLineOptions.getOrDefault(true, Collections.emptyList());
-            clashingOptions.forEach(it -> LOGGER.warn("Built-in option '{}' in task {} was disabled for clashing with another option of same name", it.getName(), task.getPath()));
-            for (OptionDescriptor optionDescriptor : validCommandLineOptions) {
+            TaskOptions taskOptions = TaskOptionsSupplier.get(task, optionReader);
+            List<OptionDescriptor> commandLineOptions = taskOptions.getAll();
+            for (OptionDescriptor optionDescriptor : commandLineOptions) {
                 String optionName = optionDescriptor.getName();
                 org.gradle.cli.CommandLineOption option = parser.option(optionName);
                 option.hasDescription(optionDescriptor.getDescription());
                 option.hasArgument(optionDescriptor.getArgumentType());
             }
+            allowOneOfEach(parser, taskOptions.getMutuallyExclusive());
```

Comment:
💭 Instead of the `CommandLineTaskConfigurer` calling this method on `taskOptions` and then the `allowOneOfEach` method processing the options to add them to the `CommandLineParser`, could the `TaskOptions` class be given a `void addMutualExclusions(CommandLineParser parser)` and encapsulate this logic there?  This might make this `CommandLineTaskConfigurer` class slightly more cohesive.

Reply:
Thanks for the suggestion, changed it. This is now similar to `BuildOptions`, which call `parser.allowOneOf` in their `configure` method.